independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/Constants.java

```diff
@@ -20,4 +20,6 @@ public interface Constants {
     String JSON = "json";
 
     String LAST_UPDATED = "last-updated";
+
+    String OFFERINGS = "offerings";
```

Comment:
Why do we care about the hardcoded key? Could it not be any key in metadata?

Reply:
For the same reason we have other "keys" for other values.

Reply:
How do you imagine it working otherwise?

Reply:
That we don't hard code the key and make it wherever users put in their config?

Reply:
It would still have to be under some key, which one?

Reply:
the key specified in the registry config:

```
registries:
- registry.acme.org:
    offerings: <---
    - "core"
    - "foo"
```

if you use that it does not need hardcoding in here does it?

Reply:
How would I check for those offerings otherwise? Are you suggesting to use `"offerings"` instead everywhere `OFFERINGS` is used?

Reply:
No. I'm suggesting no hardcoding. 

Just lookup using the keys specified in the config. 

So things like this would be legal:

```
registries:
- registry.acme.org:
    offerings: 
    - "core"
    - "foo"
    support:
    - "stable"

```

Or other combinations.

Reply:
How is this different? There is nothing that prevents this config. Are you assuming that everything under `registry` is going to be a searchable key in the extension catalog?

Reply:
Just in case, here are various registry config options that are currently supported https://github.com/quarkusio/quarkus/tree/main/independent-projects/tools/registry-client/src/test/resources/devtools-config

Reply:
> How is this different? There is nothing that prevents this config. Are you assuming that everything under `registry` is going to be a searchable key in the extension catalog?

That was my thinking yes. As we still don't have clear picture on what criteria makes sense to use.

Reply:
> Just in case, here are various registry config options that are currently supported https://github.com/quarkusio/quarkus/tree/main/independent-projects/tools/registry-client/src/test/resources/devtools-config

Yeah - so seems like a limit or filter top level would make sense which then can specify the key/values to match?

Reply:
Worth mentioning that you can have other keys under `registry` which are not really used for filtering purposes: 

eg. 

```yaml
debug: true
registries:
   - registry.quarkus.io
      maven:
        repository:
          url: http://localhost:8080/maven
      platforms:
        extension-catalogs-included: true
   - registry.quarkus.redhat.com:
       maven:
         repository:
           url: https://registry.quarkus.preprod.redhat.com/maven
```

Reply:
> Yeah - so seems like a limit or filter top level would make sense which then can specify the key/values to match?

Maybe a `filter-by` entry in the top-level would work here?

Reply:
> > Just in case, here are various registry config options that are currently supported https://github.com/quarkusio/quarkus/tree/main/independent-projects/tools/registry-client/src/test/resources/devtools-config
> 
> Yeah - so seems like a limit or filter top level would make sense which then can specify the key/values to match?

I don't see why we would assume that any option we would introduce later be used for this kind of filtering. There will be hardcoded handling of one type of options or the other.

Reply:
> > Yeah - so seems like a limit or filter top level would make sense which then can specify the key/values to match?
> 
> Maybe a `filter-by` entry in the top-level would work here?

That would be a good idea if we had multiple keys for filtering. Until this change we had none. If you have specific use-cases in mind, please describe and let's consider them.

Reply:
> > How is this different? There is nothing that prevents this config. Are you assuming that everything under `registry` is going to be a searchable key in the extension catalog?
> 
> That was my thinking yes. As we still don't have clear picture on what criteria makes sense to use.

That's a wrong assumption about searchable keys. Let's agree on the criteria.

Reply:
> > > Yeah - so seems like a limit or filter top level would make sense which then can specify the key/values to match?
> > 
> > 
> > Maybe a `filter-by` entry in the top-level would work here?
> 
> That would be a good idea if we had multiple keys for filtering. Until this change we had none. If you have specific use-cases in mind, please describe and let's consider them.

Let's clarify this further. Anything under `filter-by` would be used as a filtering key and those keys would be handled using the same principle. That raises a question why would we have different keys for the same type of filtering? Could you give an example where this would be actually useful?

Reply:
Redhat-support
Support
"Offering" 

Being able to express only include redhat supported bits. Exclude preview. Include both rhbq and camel. 

Like, why would we *only* limit this to "offering" ?

Reply:
Could you give an actual example where we would like to support these alternatives at the same time?

Reply:
Because, afaiu, technically, they are the same and can be handled by a single "offering" config. An example where splitting into alternative config option names would help.

Reply:
AFAIU, here is an equivalent of your suggestion applied to extension keywords:
Instead of
```
    "name" : "Agroal - Database connection pool",
    "description" : "Pool JDBC database connections (included in Hibernate ORM)",
    "metadata" : {
      "keywords" : [ "agroal", "database-connection-pool", "datasource", "jdbc" ],
```
we could have
```
    "name" : "Agroal - Database connection pool",
    "description" : "Pool JDBC database connections (included in Hibernate ORM)",
    "metadata" : {
      "searchable-keywords":
        "general-keywords" : [ "agroal", "database-connection-pool", "datasource", "jdbc" ],
        "potentially-other-keywords": [ "another-keyword" ]
```
So we wouldn't have `keywords` hardcoded but support any config option under `searchable-keywords` providing keywords, while a single config option `keywords` would be enough, e.g.
```
    "name" : "Agroal - Database connection pool",
    "description" : "Pool JDBC database connections (included in Hibernate ORM)",
    "metadata" : {
      "keywords" : [ "agroal", "database-connection-pool", "datasource", "jdbc", "another-keyword" ]
```

Reply:
> Because, afaiu, technically, they are the same and can be handled by a single "offering" config. An example where splitting into alternative config option names would help.

its the same reasoning I have on why we did not reuse upstream "status" for "supported" and why we used "redhat-support" rather than generic "support". To make this not red hat specific.

A single "offering" means the name space of possible values has to be globally agreed across all extensions and also means we can't fully manage them as we do today with "redhat-support". (at platform level we override the value; we don't merge them which would have to do if extensions were to use "offerings" afaics).

I don't follow what the point of "searchable-keywords" is for? I'm not suggesting we change any of the extensions ways of providing keywords. I'm suggesting to be able to express the filter using the existing metadata fields or at least not hardcode it to a global "offerings".

Reply:
> > Because, afaiu, technically, they are the same and can be handled by a single "offering" config. An example where splitting into alternative config option names would help.
> 
> its the same reasoning I have on why we did not reuse upstream "status" for "supported" and why we used "redhat-support" rather than generic "support". To make this not red hat specific.

Arguably, "status" and "support" are not in the same category.

> A single "offering" means the name space of possible values has to be globally agreed across all extensions and also means we can't fully manage them as we do today with "redhat-support". (at platform level we override the value; we don't merge them which would have to do if extensions were to use "offerings" afaics).

When it comes to "offerings", imo, an override would make sense, as well as having a unique offering name per provider.

> I don't follow what the point of "searchable-keywords" is for? I'm not suggesting we change any of the extensions ways of providing keywords. I'm suggesting to be able to express the filter using the existing metadata fields or at least not hardcode it to a global "offerings".

I am not suggesting "searchable-keywords" either but I also don't see a value in "filter-by" yet while it's another level of nesting in the config.

Reply:
"Arguably, "status" and "support" are not in the same category."

Yes. But my point was we did not make one key to express support we made a convention to allow others to express support in different ways. Ie. "acme-support" that would not conflict with whatever is in "redhat-support".

Reply:
Do we have an example where this support handling makes sense?
If an organization republishes an extension metadata with some support policies, does it make sense for it to keep the original support metadata? That joined support by multiple orgs at the same time.

Reply:
correct - a recent case for example with microsft and azure  extensions - they want to be able to express they support the extension but we would not want that to be mingled in with redhat's own notion of support.

Reply:
Would Azure re-publish RHBQ?

The example you gave is exactly hardcoding of a support provider-specific keys in both: json model and implementation of the tools handling those keys.
I don't see it as a good but a bad design decision.
It makes it impossible to implement non-vendor-specific tools that could display all support options.

Reply:
They know that *-support is for support and they be grouped that way. 

If they were all in one key called support you would instead have to prefix every value with I.e. redhat.

Reply:
I actually didn't realize we used property name expressions to represent categories. It would have been cleaner to have e.g. `support` as a parent and then nested support provider key.

Reply:
Yes but we Also wanted it to possible to capture in a hashmap and not have to repeatedly change the typed apis across different versions.